tools/loader/Loader.cpp

```diff
@@ -371,38 +410,278 @@ static void getModelInputs(std::vector<std::string> &inputNames,
 
-  // Get model input names and types.
+  // Get loader function.
+  Function *F = getFunction();
+
+  // Get model input names, types and options.
   std::vector<std::string> inputNames;
   std::vector<Type> inputTypes;
-  getModelInputs(inputNames, inputTypes);
+  std::vector<std::string> inputOpts;
+  getModelInputs(inputNames, inputTypes, inputOpts);
+
+  // Model input name and type references.
   std::vector<const char *> inputNameRefs;
   std::vector<TypeRef> inputTypeRefs;
   for (size_t idx = 0, e = inputNames.size(); idx < e; idx++) {
     inputNameRefs.push_back(inputNames[idx].c_str());
     inputTypeRefs.push_back(&inputTypes[idx]);
   }
 
+  // Get model output names and options.
+  std::vector<std::string> outputNames;
+  std::vector<std::string> outputOpts;
+  getModelOutputs(outputNames, outputOpts);
+
+  // Create the input subgraphs.
+  std::vector<Node *> inputGraphs(inputOpts.size(), nullptr);
+  for (size_t idx = 0, e = inputOpts.size(); idx < e; idx++) {
+
+    // Skip input graph creation if empty graph description.
+    if (inputOpts[idx].empty()) {
+      continue;
+    }
+
+    // Create new input placeholder using the given name and type.
+    Placeholder *newInputPlaceholder = F->getParent()->createPlaceholder(
+        inputTypeRefs[idx], inputNameRefs[idx],
+        /* isTrainable */ false);
+
+    // Create model input subgraph.
+    inputGraphs[idx] =
+        createSimpleGraphFromString(F, newInputPlaceholder, inputOpts[idx]);
+
+    // Update the type used to instatiate the model for this input using the
+    // type of this subgraph output.
+    inputTypeRefs[idx] = inputGraphs[idx]->getType(0);
+  }
+
   // Load the model based on the model format.
   std::unique_ptr<ProtobufLoader> protoLoader;
   if (!getCaffe2NetDescFilename().empty()) {
     // For Caffe2 format the input placeholder names/types must be provided
     // explicitly (mandatory).
-    protoLoader.reset(new Caffe2ModelLoader(
-        getCaffe2NetDescFilename(), getCaffe2NetWeightFilename(), inputNameRefs,
-        inputTypeRefs, *getFunction()));
+    protoLoader.reset(new Caffe2ModelLoader(getCaffe2NetDescFilename(),
+                                            getCaffe2NetWeightFilename(),
+                                            inputNameRefs, inputTypeRefs, *F));
   } else {
     // For ONNX format the input placeholders names/types can be optionally
     // provided but is not mandatory. If not provided (the arrays are empty)
     // they are derived automatically. One might want to provide explicitly
     // the input placeholder types in order to override the placeholder sizes
     // (one such example is the batch size).
     protoLoader.reset(new ONNXModelLoader(getOnnxModelFilename(), inputNameRefs,
-                                          inputTypeRefs, *getFunction()));
+                                          inputTypeRefs, *F));
+  }
+
+  // Link the input subgraphs.
```

Comment:
Same here -- it'd be great to move this logic to a local static helper to separate the logic and make it more readable/understandable.

Reply:
The procedure of adding pre-processing can only have two parts:
- before loading the model with the proto loader: create the pre-processing subgraphs and get the subgraphs output Types in order to use them as model input types
- after loading the model with the proto loader: link the pre-processing subgraphs to the model (delete placeholders and create new ones, etc).
Do you want to create two functions or keep as is?

Reply:
Yeah I'd prefer to create two functions. Thanks!